### PR TITLE
feat: calculate financial health

### PR DESCRIPTION
**Pull request by Ensar025**

### CODE REVIEW

File: packages/financial-health-calculator/src/scrapers/npm.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import fetch from 'node-fetch';
+
```
Comment (vanhoofmaarten) [inline packages/financial-health-calculator/src/scrapers/npm.ts:2]: Please remove unnecessary new rule.

File: packages/financial-health-calculator/src/data-fetchers/npm.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  let fundingUrl = '';
+
+  if (isFunding(projectData.funding)) {
+    fundingUrl = projectData.funding.url;
+  }
+
+  return {
+    fundingUrl,
+    licenseIdentifier: scrapeResult.licenseIdentifier,
```
Comment (vanhoofmaarten) [inline packages/financial-health-calculator/src/data-fetchers/npm.ts:23]: Aren't the `licenseIdentifier` and `gitHubRepositoryIdentifier` available from the `projectData`? All information visible on NPM should come from the published `package.json`

(I feel that we already had this conversation, but could you refresh my memory?)

File: packages/financial-health-calculator/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  if (!npmData.gitHubRepositoryIdentifier) {
+    throw new Error(
+      `No GitHub repository found for project: ${packageName}. Please check if their NPM page is linked to a GitHub repository.`,
+    );
+  }
+
+  const openCollectiveData = await fetchOpenCollectiveData(packageName);
+  const gitHubData = await fetchGitHubData(npmData.gitHubRepositoryIdentifier);
+
+  const acceptanceOfFundingScore = calculateAcceptanceOfFundingScore(packageName, npmData.fundingUrl, '');
```
Comment (vanhoofmaarten) [inline packages/financial-health-calculator/src/index.ts:27]: Why is the 3rd argument an empty string? If the `githubFundingUrl` isn't used, can we then remove it?

File: packages/financial-health-calculator/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    financialRoadmapScore.weightedScore +
+    organisationalActivityScore.weightedScore;
+
+  return {
+    sustainabilityScore,
+    acceptanceOfFundingScore,
+    licenseTypeScore,
+    financialRoadmapScore,
+    organisationalActivityScore,
+    finalScore: (totalScore / MAX_FINANCIAL_HEALTH_SCORE) * 100,
```
Comment (vanhoofmaarten) [inline packages/financial-health-calculator/src/index.ts:59]: Can you also place this in a separate defined `const`? Just to stick to the pattern.

File: packages/financial-health-calculator/src/calculator/acceptance-of-funding.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import type { ScoreCalculationResult } from './../types/score/score-calculation-result';
+
+export const calculateAcceptanceOfFundingScore = (
+  packageName: string,
+  npmFundingUrl: string,
+  githubFundingUrl: string,
+): ScoreCalculationResult => {
+  const weight = 0.6;
```
Comment (vanhoofmaarten) [inline packages/financial-health-calculator/src/calculator/acceptance-of-funding.ts:8]: Can we refactor these weights to be values of a constants object in the constants file?

```js
const WEIGHTS = {
  LICENSE_TYPE: 0.4,
  ORGANISATIONAL_ACTIVITY: 0.2,
  // ...
}
```

**Review by vanhoofmaarten — CHANGES_REQUESTED** [review]
@Ensar025 Thank you for this.

Had a first run through. Looking good! Already have some feedback. Will go over it again tomorrow. As this is the core of the app, let's get the details in right.

**Comment by vanhoofmaarten** [comment]
1. Can you create the serverless function that is responsible for getting the financial health?
2. Can you move the tests from `src/__tests__` to the root of the package? If we are separating the tests to a different folder, it as well can live outside the `src` folder.

File: packages/financial-health-calculator/src/index.ts (hunk 3 above)
Comment (Ensar025) [reply packages/financial-health-calculator/src/index.ts:27]: I actually forgot to add scrape the funding URL from GitHub. This is a TODO. I think I put in an empty string just to check out if everything else works and then forgot

File: packages/financial-health-calculator/src/secrets.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export const OPEN_COLLECTIVE_API_KEY = 'd012b7336617a21c15650b05e1dec4b0d02e1215';
```
[the remark this answers is not in the document]
Comment (Ensar025) [reply packages/financial-health-calculator/src/secrets.ts]: Resolved in a938286a160db3df98eaecabca38c96d4714f8fc

File: packages/financial-health-calculator/src/calculator/acceptance-of-funding.ts (hunk 5 above)
Comment (Ensar025) [reply packages/financial-health-calculator/src/calculator/acceptance-of-funding.ts]: Resolved in 46a93d61fa8534f67d9b794fc06bd5620ec9225f

File: packages/financial-health-calculator/src/calculator/license-type.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { spdxLicenseList } from '../data-files/spdx-licenses.js';
```
[the remark this answers is not in the document]
Comment (Ensar025) [reply packages/financial-health-calculator/src/calculator/license-type.ts]: Resolved in e1b8121ba189f53e13fbdecbb3e712327a980422

**Comment by Ensar025** [comment]
> 1. Can you create the serverless function that is responsible for getting the financial health?
> 2. Can you move the tests from `src/__tests__` to the root of the package? If we are separating the tests to a different folder, it as well can live outside the `src` folder.

1. I was thinking of doing that in a separate PR, can do it also here though
2. Alright, will do

File: packages/financial-health-calculator/src/data-fetchers/npm.ts (hunk 2 above)
Comment (Ensar025) [reply packages/financial-health-calculator/src/data-fetchers/npm.ts:23]: If you include the fullMetadata option it also fetches the funding URL and license type:
```js
{
  name: 'eslint',
  version: '8.16.0',
  author: {
    name: 'Nicholas C. Zakas',
    email: 'nicholas+npm@nczconsulting.com'
  },
  description: 'An AST-based pattern checker for JavaScript.',
  bin: { eslint: 'bin/eslint.js' },
  main: './lib/api.js',
  exports: {
    './package.json': './package.json',
    '.': './lib/api.js',
    './use-at-your-own-risk': './lib/unsupported-api.js'
  },
  scripts: {
    ...scripts
  },
  gitHooks: { 'pre-commit': 'lint-staged' },
  'lint-staged': {
    '*.js': 'eslint --fix',
    '*.md': 'markdownlint --fix',
    'docs/src/rules/*.md': [
      'node tools/fetch-docs-links.js',
      'git add docs/src/_data/further_reading_links.json'
    ]
  },
  repository: { type: 'git', url: 'git+https://github.com/eslint/eslint.git' },
  funding: { url: 'https://opencollective.com/eslint' },
  homepage: 'https://eslint.org',
  bugs: { url: 'https://github.com/eslint/eslint/issues/' },
  dependencies: {
    ...deps
  },
  devDependencies: {
    ...devDeps
  },
  keywords: [ 'ast', 'lint', 'javascript', 'ecmascript', 'espree' ],
  license: 'MIT',
  engines: { node: '^12.22.0 || ^14.17.0 || >=16.0.0' },
  gitHead: '901ce0f1e32ea1e9e10ce4d8b37c0d750007a3c5',
  _id: 'eslint@8.16.0',
  _nodeVersion: '16.13.1',
  _npmVersion: '8.1.2',
  dist: {
    integrity: ...key
  },
  _npmUser: { name: 'eslintbot', email: 'nicholas+eslint@nczconsulting.com' },
  directories: {},
  maintainers: [
    { name: 'openjsfoundation', email: 'npm@openjsf.org' },
    { name: 'eslintbot', email: 'nicholas+eslint@nczconsulting.com' },
    { name: 'nzakas', email: 'nicholas@nczconsulting.com' },
    { name: 'ivolodin', email: 'ivolodin@gmail.com' }
  ],
  _npmOperationalInternal: {
    host: 's3://npm-registry-packages',
    tmp: 'tmp/eslint_8.16.0_1653086291946_0.7993754545335081'
  },
  _hasShrinkwrap: false,
  _resolved: 'https://registry.npmjs.org/eslint/-/eslint-8.16.0.tgz',
  _from: 'eslint@',
  _integrity: 'sha512-MBndsoXY/PeVTDJeWsYj7kLZ5hQpJOfMYLsF6LicLHQWbRDG19lK5jOix4DPl8yY4SUFcE3txy86OzFLWT+yoA==',
  _signatures: [...signatures
  ]
}
`

File: packages/financial-health-calculator/src/index.ts (hunk 3 above)
Comment (vanhoofmaarten) [reply packages/financial-health-calculator/src/index.ts:27]: Ok. Let's pick this up in a separate PR for optimization.

File: packages/financial-health-calculator/src/data-fetchers/npm.ts (hunk 2 above)
Comment (vanhoofmaarten) [reply packages/financial-health-calculator/src/data-fetchers/npm.ts:23]: I also see a `repository` property of which we could derive the repository page URL.

File: packages/financial-health-calculator/src/scrapers/npm.ts (hunk 1 above)
Comment (vanhoofmaarten) [reply packages/financial-health-calculator/src/scrapers/npm.ts:2]: Oh, I wasn't referring to `node-fetch`, just the empty line on rule 2.

Comment (Ensar025) [reply packages/financial-health-calculator/src/scrapers/npm.ts:2]: Oh :D

File: packages/financial-health-calculator/src/data-fetchers/npm.ts (hunk 2 above)
Comment (Ensar025) [reply packages/financial-health-calculator/src/data-fetchers/npm.ts:23]: Yes, I think we can

Comment (vanhoofmaarten) [reply packages/financial-health-calculator/src/data-fetchers/npm.ts:23]: Do it

Do it.

File: packages/financial-health-calculator/src/scrapers/npm.ts (hunk 1 above)
Comment (Ensar025) [reply packages/financial-health-calculator/src/scrapers/npm.ts]: NPM scraper has been removed completely

File: packages/financial-health-calculator/src/data-fetchers/npm.ts (hunk 2 above)
Comment (Ensar025) [reply packages/financial-health-calculator/src/data-fetchers/npm.ts]: NPM scraper is removed

File: packages/backend/financial-health-calculator/index.mts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import { calculateFinancialHealthScore } from '@dependency-health-monitor/financial-health-calculator';
+import type { AzureFunction, Context, HttpRequest } from '@azure/functions';
+
+const httpTrigger: AzureFunction = async function (context: Context, req: HttpRequest): Promise<void> {
```
Comment (vanhoofmaarten) [inline packages/backend/financial-health-calculator/index.mts:4]: Can you add the following line?
`context.res.type('application/json');`

**Review by vanhoofmaarten — CHANGES_REQUESTED** [review]
Hi @Ensar025. Looking good. Have added some questions and some minor change requests.

File: packages/financial-health-calculator/package.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  "dependencies": {
+    "dotenv": "^16.0.1",
+    "node-html-parser": "^5.3.3",
+    "p-map": "^5.4.0",
+    "pacote": "^13.4.1",
+    "spdx-license-list": "^6.5.0"
+  },
+  "devDependencies": {
+    "@types/pacote": "^11.1.3",
+    "c8": "^7.11.3",
```
[the remark this answers is not in the document]
Comment (Ensar025) [reply packages/financial-health-calculator/package.json:23]: Not anymore no. c8 is required for dotenv, but we don't need env vars anymore. Will remove

File: .gitignore [hunk 10]
```diff
@@ -34,4 +34,7 @@ build
 .next
 
 # Azure Functions
-local.settings.json
\ No newline at end of file
+local.settings.json
+
+# Financial health calculation
```
[the remark this answers is not in the document]
Comment (Ensar025) [reply .gitignore]: Resolved in 1d26040078727b6caa2f18257ec7fb11cf1d0cca

File: packages/financial-health-calculator/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ensar025) [reply packages/financial-health-calculator/package.json]: Resolved in c94ef1eca8114b99c641e19ec28f478736dc1537

File: packages/backend/npm-dependency-fetcher/index.mts [hunk 11]
```diff
@@ -2,14 +2,14 @@ import { fetchDependenciesForPackageInDepth } from '@dependency-health-monitor/n
 import type { AzureFunction, Context, HttpRequest } from '@azure/functions';
 
 const httpTrigger: AzureFunction = async function (context: Context, req: HttpRequest): Promise<void> {
```
[the remark this answers is not in the document]
Comment (Ensar025) [reply packages/backend/npm-dependency-fetcher/index.mts:4]: Resolved in d8446cb24a86f01371f0dbb722b68ad87f93e311

File: packages/backend/financial-health-calculator/index.mts (hunk 8 above)
Comment (Ensar025) [reply packages/backend/financial-health-calculator/index.mts:4]: Resolved in d8446cb24a86f01371f0dbb722b68ad87f93e311

File: .eslintrc.cjs [hunk 12]
```diff
@@ -6,6 +6,7 @@ module.exports = {
   extends: ['eslint:recommended', '@vue/eslint-config-typescript/recommended', '@vue/eslint-config-prettier'],
   env: {
     'vue/setup-compiler-macros': true,
+    node: true,
```
[the remark this answers is not in the document]
Comment (Ensar025) [reply .eslintrc.cjs]: Actually, this configuration is not needed anymore. I removed it in 91465e2444d72ff31bfc23721e1eb6c87b4534b3

[the remark this answers is not in the document]
Comment (Ensar025) [reply .eslintrc.cjs]: It was needed when using environment variables, but we don't use environment variables anymore

**Review by vanhoofmaarten — APPROVED** [review]

**Pull request merged**